src/imex.rs

```diff
@@ -106,6 +107,178 @@ pub async fn imex(
+    let recv_db = recv_dir.path().join("db");
+    let port = 9991;
+
+    let receiver = Receiver::new(port, &recv_db)
+        .await
+        .context("failed to create sender")?;
+    let mut receiver_transfer = receiver
+        .transfer_from_ticket(&ticket)
+        .await
+        .context("failed to read transfer")?;
+    let data = receiver_transfer.recv().await?;
+    let mut progress = receiver_transfer.progress()?;
+
+    context.sql.config_cache.write().await.clear();
+
+    // progress report
+    let ctx = context.clone();
+    let progress_task = tokio::spawn(async move {
+        let mut last_progress = 0;
+        while let Some(ev) = progress.next().await {
+            match ev {
+                Ok(iroh_share::ProgressEvent::Piece { index, total }) => {
+                    let progress = 1000 * index / total;
+                    if progress != last_progress && progress < 1000 {
+                        ctx.emit_event(EventType::ImexProgress(progress));
+                        last_progress = progress;
+                    }
+                }
+                Err(err) => {
+                    error!(ctx, "IMEX receive backup failed to complete: {}", err);
+                    ctx.emit_event(EventType::ImexProgress(0));
+                }
+            }
+        }
+    });
+
+    let out = context.get_blobdir();
+
+    let mut read_dir_stream = data
+        .read_dir()?
+        .ok_or_else(|| anyhow::anyhow!("unexpected data structure"))?;
+
+    while let Some(link) = read_dir_stream.next().await {
+        let link = link?;
+        let file_content = data.read_file(&link).await?;
+        let name = link.name.unwrap_or_default();
+        let path = out.join(&name);
+
+        let mut file = tokio::fs::File::create(&path)
+            .await
+            .with_context(|| format!("create file: {}", path.display()))?;
+        let mut content = file_content.pretty()?;
+        tokio::io::copy(&mut content, &mut file)
+            .await
+            .context("copy")?;
+        file.flush().await?;
+
+        if name == DBFILE_BACKUP_NAME {
+            let unpacked_database = context.get_blobdir().join(DBFILE_BACKUP_NAME);
+            context
+                .sql
+                .import(&unpacked_database, passphrase.clone())
+                .await
+                .context("cannot import unpacked database")?;
+            fs::remove_file(unpacked_database)
+                .await
+                .context("cannot remove unpacked database")?;
+        } else {
+            // nothing to do, unpacked directly into the blobs dir
+        }
+    }
+
+    progress_task.await?;
+    receiver_transfer.finish().await?;
+
+    info!(context, "Received all data, written to: {}", out.display());
+
+    Ok(())
+}
+
+pub async fn send_backup(
+    context: &Context,
+    path: &Path,
+    passphrase: Option<String>,
+) -> Result<iroh_share::SenderTransfer> {
+    let cancel = context.alloc_ongoing().await?;
```

Comment:
I assume sending a backup was purposefully made an ongoing process and this is the desired UI experience?  Like you shouldn't be able to do anything else while sending a backup?

Reply:
correct